File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    public static BufferedReader br;
    public static StringTokenizer st;
    public static int totalNode, node;
    public static ArrayList<Integer>[] graph;
    public static boolean[] visited;
    public static int[] parentsNode;
    public static Queue<Integer> queue;


    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        totalNode = Integer.parseInt(br.readLine());
        graph = new ArrayList[totalNode+1];
        parentsNode = new int[totalNode+1];
        queue = new LinkedList<>();
        visited = new boolean[totalNode+1];
        for(int idx = 0; idx < totalNode+1; idx++){
            graph[idx] = new ArrayList<>();
        }
        // 그래프 초기화
        for(int cnt = 0; cnt < totalNode-1; cnt++){
            st = new StringTokenizer(br.readLine());
            int node1 = Integer.parseInt(st.nextToken());
            int node2 = Integer.parseInt(st.nextToken());
            graph[node1].add(node2); graph[node2].add(node1);
        }
        BFS();
        for(int node = 2; node < parentsNode.length; node++){
            System.out.println(parentsNode[node]);
        }

    }

    static void BFS(){
        queue.add(1);
        visited[1] = true;
        while(!queue.isEmpty()){
            int node = queue.poll();

            for (Integer v : graph[node]) {
                if (!visited[v]){
                    queue.add(v);
                    visited[v] = true;
                    parentsNode[v] = node;
                }
            }
        }
    }
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
sys.setrecursionlimit(10**5)
input = sys.stdin.readline

n = int(input())
graph = [[] for _ in range(n+1)]
visited = [0 for _ in range(n+1)]

for i in range(n-1):
    v1,v2 = map(int,input().split())
    graph[v1].append(v2)
    graph[v2].append(v1)

def dfs(v1):
    for v2 in graph[v1]:
        if not visited[v2]:
            visited[v2] = v1
            dfs(v2)

for i in range(1,n+1):
    if not visited[i]:
        visited[i] = i
        dfs(i)

for v in visited[2:]:
    print(v)